File: lib/middleware/proxy.js
/*
 * Copyright 2014 Telefonica Investigación y Desarrollo, S.A.U
 *
 * This file is part of fiware-pep-steelskin
 *
 * fiware-pep-steelskin is free software: you can redistribute it and/or
 * modify it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the License,
 * or (at your option) any later version.
 *
 * fiware-pep-steelskin is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public
 * License along with fiware-pep-steelskin.
 * If not, seehttp://www.gnu.org/licenses/.
 *
 * For those usages not covered by the GNU Affero General Public License
 * please contact with::[iot_support@tid.es]
 */

'use strict';

var config = require('../../config'),
    errors = require('../errors'),
    request = require('../request-shim'),
    constants = require('../constants'),
    validationHeaders = [
        'fiware-service',
        'fiware-servicepath',
        'x-auth-token'
    ],
    authorizationHeaders = [
        'x-auth-token'
    ],
    winston = require('winston'),
    logger = require('logops'),
    configAccessMatch = require('../../configAccessMatch.js').configAccessMatch,
    accessLogger;

const fs = require('fs');
const configAccessMatchFilePath = './configAccessMatch.js';

function requireUncached(module) {
    delete require.cache[require.resolve(module)];
    return require(module);
}

function watchConfigAccessMatchFile() {
    fs.watch(configAccessMatchFilePath, (event, filename) => {
        logger.info('watchConfigAccessMatchFile changed by %s detected in file %s', event, filename);
        try {
            configAccessMatch = requireUncached('../../configAccessMatch.js').configAccessMatch;
            logger.debug('reloaded configAccessMatch %j', configAccessMatch);
        } catch (err) {
            logger.error('Error %s reloading module: %s ', err, filename);
        }
    });
}

/**
 * Middleware to extract the organization data from the request.
 *
 * @param {Object} req           Incoming request.
 * @param {Object} res           Outgoing response.
 * @param {Function} next        Call to the next middleware in the chain.
 */
function extractOrganization(req, res, next) {
    const logger = req.logger;
    if (req.headers[constants.ORGANIZATION_HEADER]) {
        req.organization = req.headers[constants.ORGANIZATION_HEADER];
        req.service = req.headers[constants.ORGANIZATION_HEADER];
        req.subService = req.headers[constants.PATH_HEADER];
        next();
    } else {
        logger.error('[PROXY-GEN-001] Organization headers not found');
        next(new errors.OrganizationNotFound());
    }
}

/**
 * Middleware to extract the user data (usually a token) from the request.
 *
 * @param {Object} req           Incoming request.
 * @param {Object} res           Outgoing response.
 * @param {Function} next        Call to the next middleware in the chain.
 */
function extractUserId(req, res, next) {
    const logger = req.logger;
    if (req.headers[constants.AUTHORIZATION_HEADER]) {
        req.userId = req.headers[constants.AUTHORIZATION_HEADER];
        next();
    } else {
        logger.error('[PROXY-GEN-002] User ID headers not found');
        next(new errors.UserNotFound());
    }
}

/**
 * Generates the FRN to identify the target resource of the request. The FRN is composed of information from the
 * service and subservice headers and some configuration information.
 *
 * @param {Object} req           Incoming request.
 * @param {Object} res           Outgoing response.
 * @param {Function} next        Invokes the next middleware in the chain.
 */
function generateFRN(req, res, next) {
    const logger = req.logger;
    var frn = config.resourceNamePrefix + config.componentName + ':';

    if (req.organization) {
        frn += req.organization + ':';
    } else {
        frn += ':';
    }

    if (req.headers[constants.PATH_HEADER]) {
        frn += req.headers[constants.PATH_HEADER] + ':';
    } else {
        frn += ':';
    }

    if (req.resourceUrl) {
        frn += req.resourceUrl + ':';
    } else {
        frn += ':';
    }

    if (req.entityType) {
        frn += req.entityType;
    } else {
        frn += ':';
    }
    logger.debug('Generated FRN: ', frn);
    req.frn = frn;

    next();
}

/**
 * Redirects the incoming request to the proxied host.
 * read to guess its type.
 *
 * @param {Object} req           Incoming request.
 * @param {Object} res           Outgoing response.
 * @param {Function} next        Invokes the next middleware in the chain.
 */
function sendRequest(req, res, next) {
    const logger = req.logger;
    var options = {
        uri: 'http://' + config.resource.original.host + ':' + config.resource.original.port + req.path,
        qs: req.query,
        method: req.method,
        headers: req.headers
    };

    if (!options.headers[constants.X_FORWARDED_FOR_HEADER] && req.connection.remoteAddress) {
        options.headers[constants.X_FORWARDED_FOR_HEADER] = req.connection.remoteAddress;
    }
    if (!options.headers[constants.CORRELATOR_HEADER] && req.corr) {
        options.headers[constants.CORRELATOR_HEADER] = req.corr;
    }

    if (req.is('*/json')) {
        options.body = JSON.stringify(req.body);
    } else {
        options.body = req.rawBody;
    }

    delete options.headers['content-length'];
    options.headers.connection = 'close';

    res.oldWriteHead = res.writeHead;
    res.writeHead = function(statusCode, reasonPhrase, headers) {
        if (res._headers['transfer-encoding']) {
            delete res._headers['transfer-encoding'];
        }

        res.oldWriteHead(statusCode, reasonPhrase, headers);
    };

    logger.debug('Forwarding request:\n\n%j\n', options);

    req.fwdResponse = request(options).on('error', function handleConnectionError(e) {
        logger.error('Error forwarding the request to target proxy: %s', e.message);

        if (config.dieOnRedirectError) {
            logger.fatal('[PROXY-FATAL-001] Configured to die upon error in a redirection. Stopping process.');

            process.exit(-1);
        } else {
            next(new errors.TargetServerError(e.message));
        }
    });
    next();
}

/**
 * Check here MATCH file patterns
 *
 * @param {Object} req           Incoming request.
 * @param {String} accessMsg     Incoming accessMsg
 * @return {String}              String message corresponding with accessMsg and found matches
 */
function checkAccessMatches(req, accessMsg) {
    if (req.userName in configAccessMatch.users ) {
        accessMsg += ' MATCHED USER ' + req.userName;
    }
    for (var header of configAccessMatch.headers) {
        var headerName = Object.keys(header)[0];
        if (Object.keys(req.headers).includes(headerName)) {
            if (req.headers[headerName] === header[headerName]) {
                accessMsg += ' MATCHED HEADER ' + headerName + ' ' + header[headerName];
            }
        }
    }
    for (var subpath of configAccessMatch.subpaths) {
        if (req.path.includes(subpath)) {
            accessMsg += ' MATCHED SUBPATH ' + subpath;
        }
    }
    for (var subquery of configAccessMatch.subqueries) {
        if (JSON.stringify(req.query).includes(subquery)) {
            accessMsg += ' MATCHED SUBQUERY ' + subquery;
        }
    }
    for (var text of configAccessMatch.body) {
        if (JSON.stringify(req.body).includes(text)) {
            accessMsg += ' MATCHED BODY ' + text;
        }
    }
    return accessMsg;
}


/**
 * Account Log
 * read to guess its type.
 *
 * @param {Object} req           Incoming request.
 * @param {Object} res           Outgoing response.
 * @param {Function} next        Invokes the next middleware in the chain.
 */
function accountInfo(req, res, next) {
    if ('fwdResponse' in req) {
        if (!accessLogger) {
            accessLogger = new winston.Logger({
                level: 'info',
                transports: [
                    new(winston.transports.File)({
                        filename: config.access.accountFile,
                        json: false,
                        timestamp: false,
                        showLevel: false
                    })
                ]
            });
        }
        req.fwdResponse = req.fwdResponse.on('response', function(res) {
            var accessMsgOrig = 'Right Attempt';
            var accessMsg = accessMsgOrig;
            if (! ['wrong'].includes(config.access.accountMode) ) {
                accessMsg = checkAccessMatches(req, accessMsgOrig);
            }
            if ( ['all'].includes(config.access.accountMode) ||
                 (['matched'].includes(config.access.accountMode) &&
                  accessMsgOrig.length < accessMsg.length) ){

                accessLogger.info(accessMsg +
                                  ' | ResponseStatus=' + req.fwdResponse.response.statusCode +
                                  ' | Token=' + req.headers['x-auth-token'] +
                                  ' | Origin=' + (req.ip || req.connection.remoteAddress) +
                                  ' | UserId=' + req.userId +
                                  ' | UserName=' + req.userName +
                                  ' | ServiceId=' + req.serviceId +
                                  ' | Service=' + req.service +
                                  ' | SubServiceId=' + req.subserviceId +
                                  ' | SubService=' + req.subService +
                                  ' | Action=' + req.action +
                                  ' | Path=' + req.path +
                                  ' | Query=' + JSON.stringify(req.query) +
                                  ' | Body=' + JSON.stringify(req.body).slice(0, 100) + // not all body
                                  ' | Date=' + new Date().toJSON());
            }
        });
    }
    next();
}

/**
 * Invoque request and get response. The request is not read with a pipe, as it has been completely
 * read to guess its type.
 *
 * @param {Object} req           Incoming request.
 * @param {Object} res           Outgoing response.
 * @param {Function} next        Invokes the next middleware in the chain.
 */
function sendResponse(req, res, next) {
    req.fwdResponse.pipe(res);
}

/**
 * Generates a middleware that checks for the pressence of the mandatory headers passed as a parameter, returning a
 * MISSING_HEADERS error if any one is not found.
 *
 * @param {Array} mandatoryHeaders      List of headers to check.
 * @return {Function}                  An express middleware that checks for the presence of the headers.
 */
function checkMandatoryHeaders(mandatoryHeaders) {
    return function(req, res, next) {
        var missing = [];

        for (var i = 0; i < mandatoryHeaders.length; i++) {
            if (!req.headers[mandatoryHeaders[i]] || req.headers[mandatoryHeaders[i]].trim() === '') {
                missing.push(mandatoryHeaders[i]);
            }
        }

        if (missing.length !== 0) {
            next(new errors.MissingHeaders(JSON.stringify(missing)));
        } else {
            next();
        }
    };
}

exports.generateFRN = generateFRN;
exports.extractUserId = extractUserId;
exports.extractOrganization = extractOrganization;
exports.sendRequest = sendRequest;
exports.sendResponse = sendResponse;
exports.accountInfo = accountInfo;
exports.checkMandatoryHeaders = checkMandatoryHeaders(validationHeaders);
exports.checkAuthorizationHeader = checkMandatoryHeaders(authorizationHeaders);
exports.watchConfigAccessMatchFile = watchConfigAccessMatchFile;
exports.checkAccessMatches = checkAccessMatches;

File: test/acceptance/tools/mocks/proxy.py
# -*- coding: utf-8 -*-
"""
Copyright 2014 Telefonica Investigación y Desarrollo, S.A.U

This file is part of fiware-orion-pep

fiware-orion-pep is free software: you can redistribute it and/or
modify it under the terms of the GNU Affero General Public License as
published by the Free Software Foundation, either version 3 of the License,
or (at your option) any later version.

fiware-orion-pep is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
See the GNU Affero General Public License for more details.

You should have received a copy of the GNU Affero General Public
License along with fiware-orion-pep.
If not, seehttp://www.gnu.org/licenses/.

For those usages not covered by the GNU Affero General Public License
please contact with::[iot_support@tid.es]
"""
__author__ = 'Jon Calderin Goñi <jon.caldering@gmail.com>'

from flask import Flask, request, Response
import requests
import sys
import collections
from iotqautils.iotqaLogger import get_logger

app = Flask(__name__)


log = get_logger('proxy')

requested = ''
last_path = ''
history = []

def convert(data):
    if isinstance(data, basestring):
        return str(data)
    elif isinstance(data, collections.Mapping):
        return dict(map(convert, data.iteritems()))
    elif isinstance(data, collections.Iterable):
        return type(data)(map(convert, data))
    else:
        return data


@app.route('/', defaults={'path': ''}, methods=['GET', 'POST', 'UPDATE', 'DELETE', 'PATCH', 'HEAD', 'OPTIONS'])
@app.route('/<path:path>', methods=['GET', 'POST', 'UPDATE', 'DELETE', 'PATCH', 'HEAD', 'OPTIONS'])
def proxy(path):
    global requested
    global last_path
    global history
    log.debug('********************************* Entring in proxy***********************************************************')
    if path == 'last_path':
        ret_last_path = last_path
        last_path = ''
        return ret_last_path
    elif path == 'history':
        return str(history)
    elif path == 'reset_history':
        history = []
        return ''
    else:
        history.append(path)
        last_path = path
        url = request.scheme + '://%s:%s/%s' % (sys.argv[3], sys.argv[4], path)
        headers = convert(dict(request.headers))
        headers['Host'] = "{ip_dest}:{port_dest}".format(ip_dest=sys.argv[3], port_dest=sys.argv[4])
        del headers['Content-Length']
        method = request.method.lower()
        redirect = False
        params = request.args
        stream = False
        timeout = 30
        log.debug('Request: \n\n Method: %s \n\n Headers: %s \n\n Data: %s \n\n URL: %s \n\n ARGS: %s \n\n---------------------------------------------' % \
        (str(method), str(headers), str(request.data), str(url), str(params)))
        if method == 'post':
            log.debug('Sending %s headers' % method)
            r = requests.request(method, url, allow_redirects=redirect, headers=headers, params=params, stream=stream, data=request.data, timeout=timeout)
        else:
            log.debug('Sending %s headers' % method)
            r = requests.request(method, url, allow_redirects=redirect, headers=headers, params=params, stream=stream, timeout=timeout)

        headers_resp = dict(r.headers)
        if 'transfer-encoding' in headers_resp:
            del headers_resp['transfer-encoding']
        response_data = r.content
        status_code = r.status_code
        flask_response = Response(response=response_data,
                                  status=status_code,
                                  headers=headers_resp.items())
        log.debug('Response: \n\n  Headers: %s \n\n Data: %s \n\n StatusCode: %s \n\n Response: %s \n\n+++++++++++++++++++++++++++++++++++++++++++++' % \
        (headers_resp, response_data, status_code, flask_response.response))
    log.debug('################################### Exiting proxy ##############################################################\n\n\n')
    return flask_response

if __name__ == '__main__':
    if len(sys.argv) < 5:
        raise NameError('You have to indicate the host and the port of the proxy and the host and the port for the destination')
    app.run(host=str(sys.argv[1]), port=int(sys.argv[2]), debug=True)
